### ISSUE TITLE
[App Request] How would you feel about an org-mode app?

### ISSUE THREAD

**Issue by 9z0** [ransome1/sleek#160]
### Advantages of todo.txt:

- Simplicity by design. One line.
- **Text-file based** (editable, future-proof, etc).

### Disadvantages of todo.txt:
- Not everything can be reasonably done in **one line**.

### Current apps for todo.txt:
- **Desktop**: good number, although maintenance and features vary.
- **Mobile**: reasonable number.

---

### Advantages of org-mode:

- Supports an arbitrary number of lines per node/task.
- **Text-file based**.
- Can be as simple as todo.txt or deeply customized and extended.

### Disadvantages of org-mode:
- Not easily accessible outside _Emcas_ (see below), which isn't for everyone and has a steep learning curve to say the least.

### Current apps for org-mode:
- **Desktop**: Barely any apps outside the Emacs ecosystem. I looked. For a LONG time.
    - Extensions for VS Code, Atom, Sublime, etc:
        - Potential is limited by the API of the parent app.
        - Text-based support. No bars, panels, buttons, good visual manipulation, etc.
- **Mobile**: Only a couple of apps. For example Orgzly.

---

- If you look at the discussions of enthusiasts online, I think there is a real need for a good, accessible GUI to org-mode:
    - [Get started with Org mode without Emacs](https://opensource.com/article/19/1/productivity-tool-org-mode)
    - https://www.reddit.com/r/orgmode/comments/lcizik/does_orgmode_outside_of_emacs_make_sense/
    - And see the discussions on multiple Hacker News posts. It gives a good idea about how the org-mode situation is seen by many technical people. [1](https://news.ycombinator.com/item?id=16198369), [2](https://news.ycombinator.com/item?id=23299993), [3](https://news.ycombinator.com/item?id=19622019), [4](https://news.ycombinator.com/item?id=13828732), [5](https://news.ycombinator.com/item?id=10925655).

---

### My suggestion

Even with the better situation of todo.txt, **Sleek** seems to have struck a chord with users, especially considering that it is not a year-old in public yet. Many reports (of bugs) and specific requests suggest active usage rather than just "try this app and wait for maturity". It **satisfies a need that wasn't really met before**. I have been using todo.txt for years now and struggling to go beyond simple usage until Sleek.

Org-mode needs an app that enables that usage where the open, text-based format is the robust foundation instead of the whole experience. You get ease of use AND advantages of text and an open standard.

There are JS parsers and web apps for org-mode, but not a desktop app. The situation is ripe for a good one. I think an app that covers even the very basics of org-mode visually will be received with welcome

---

### TLDR

- todo.txt is good. It has apps already. But people like Sleek. Good GUI/UX.
- org-mode is good. It has no apps.
- You understand the benefit of text formats.
- You are good with JavaScript. JavaScript base for org-mode exists.
- Basic org-mode app would be good. You make **Slook**. People like **Slook**. You help people! 😃

**Comment by 9z0** [comment]
To be clear, I am NOT necessarily talking about a "pretty" UI like [this](https://lepisma.xyz/2017/10/28/ricing-org-mode/). That would be a bonus but it can be achieved with many markup languages and with apps/extension like those for VS Code.

I meant grouping and managing of bits of information/todos (tagging/filtering/searching, etc). Like what todo.txt and Sleek provide, but with the power of org-mode (multi-line notes, formatting)!

If you see [Orgzly](https://github.com/orgzly/orgzly-android/), you would understand how it can be used for anything!

[CintaNotes](http://cintanotes.com/) is the closed-source equivalent of the organizational aspect of org-mode.

**Comment by ransome1** [comment]
@9z0 , I think this is a great idea. Last year I gave org-mode a try but didn't stick to it, so I don't have a lot of experience with it. 

If there are JS parsers already, the process of getting org-mode text files displayed in sleek or a fork of sleek, should not be an issue. If sleek would be able to identify a file as being a todo.txt or org mode file, it could use different parsers to continue. So one app would be able to handle both. 

The challenge in general would be to figure out what org-mode adds on the feature side and what would need to be done to make sleek handle those org-mode features properly.

I will take a closer look at org mode soon and try to figure out, what would need to be done.

**Comment by 9z0** [comment]
Glad you liked it.

You can look into these projects (under various licenses):
https://github.com/DanielDe/org-web
https://github.com/200ok-ch/organice
https://github.com/daitangio/org-mode-parser
https://github.com/borablanca/orgmodeweb
https://github.com/fusupo/org-parse
https://github.com/orgapp/orgajs (TypeScript)
https://github.com/rasendubi/uniorg (TypeScript)
https://github.com/k2052/org-to-markdown
https://github.com/mooz/org-js (A bit old)
https://github.com/Patrickskiba/weborg
https://github.com/keenan691/org-mode-connection
https://github.com/jsorg/jsorg (old)

There are also these VS Code extensions:
https://github.com/vscode-org-mode/vscode-org-mode
https://github.com/aboukirev/vscode-orgmode
https://github.com/robaboyd/vs-org
https://github.com/crz33/vscode-markdown-agenda
I think i

**Comment by 9z0** [comment]
Glad you liked it.

You can look into these projects (under various licenses):
https://github.com/DanielDe/org-web
https://github.com/200ok-ch/organice
https://github.com/daitangio/org-mode-parser
https://github.com/borablanca/orgmodeweb
https://github.com/fusupo/org-parse
https://github.com/maxminoS/markup-parser (TypeScript)
https://github.com/orgapp/orgajs (TypeScript)
https://github.com/rasendubi/uniorg (TypeScript)
https://github.com/k2052/org-to-markdown
https://github.com/mooz/org-js (A bit old)
https://github.com/Patrickskiba/weborg
https://github.com/keenan691/org-mode-connection
https://github.com/jsorg/jsorg (old)
https://github.com/jsns/gorg-mode (TypeScript)

There are also these VS Code extensions:
https://github.com/vscode-org-mode/vscode-org-mode
https://github.com/aboukirev/vscode-orgmode
https://github.com/robaboyd/vs-org
https://github.com/crz33/vscode-markdown-agenda

---

>in sleek or a fork of sleek

I think it may be better not to use the same app for both. The formats (todo.txt and org-mode) are different enough that a single UI model (or even options) may not work for both.

Using as much as possible of the code base for both should help with maintenance, of course.

Good luck!

**Comment by zerodat** [comment]
Before returning to todo.txt a few weeks ago, I had been using orgzly and org-mode for about a year. And before that I was using todo.txt with topydo, QTodoTxt, and simpletask-android for a couple of years.  So I can compare the two systems with a fair amount of context.

The TL;DR is that I don't think sleek should try to handle both formats.  I certainly didn't use all of org-mode, nor do I claim to understand every option that it offers.  It is no doubt very useful for some folks.  But many of us just want to have the minimum amount of complexity while keeping track of all of our tasks in a simple, reliable way.   Sleek and todo.txt seem to do this well. Here are some more observations.

* Orgzly is a very nice Android app.  Attractive and works well, although I found the synchronization behavior was a little tricky.
* I didn't find a good desktop app, partly because true org-mode fans will usually use emacs where org-mode was created.  So there is a open niche for a nice GUI app, but it probably wouldn't get those emacs users to leave their favorite tool.
* one of the big todo.txt advantages is that line-oriented unix/linux shell tools can easily process tasks one per line, and you lose that simplicity with org-mode.
*  I found that I didn't need most of the extra complexity that org-mode format offers.  While it is nice to have multi-line task descriptions or notes, you still need to keep a short one-line heading description that you can recognize in a task list.  I'm willing to work around the single-line limitation of todo.txt (and it can be a LONG line!) in order to keep the simplicity.
* org-mode is intended to be a general outliner, note-taker, and Markdown alternative, as well as a task manager.  When I took notes in orgzly, I would eventually wish that I could get them out of there and into my note-taking system, which is a wiki-like web app.
* Task states like "TODO" and "DONE" are just defaults in org-mode which can be customized to anything the user wants.  Org-mode fans will build quite elaborate workflows around this with many more than two states for a task.
* There are multiple dates that can be associated with an item: SCHEDULED, DEADLINE, COMPLETED, etc. Each date can have times, time ranges, etc.  One use of org-mode is to track the time you spent working on each task for time-accounting and billing.
* One thing I did like about org-mode was the ability to have tasks with sub-tasks (a general tree of tasks).  But in practice I found that I hardly ever used this feature.  In todo.txt, the +project tag is a limited substitute for this, but with the advantage that a subtask can have multiple "parent" projects because they are just tags.  Org-mode has tags too, and I actually used those more than I used subtasks.

**Comment by 9z0** [comment]
Nice insights. I generally agree.

>I didn't find a good desktop app, partly because true org-mode fans will usually use emacs where org-mode was created. So there is a open niche for a nice GUI app, but it probably wouldn't get those emacs users to leave their favorite tool.

This is the point. I don't see them are a significant target audience. It is for us non-Emacs plebs!

>I didn't need most of the extra complexity that org-mode format offers.

I personally want the organization ability more than anything. It represents for me a **text-based relational database**!

Because it inherently supports nesting/outlining, tags (with inheritance), multilines, _custom attributes_, date/time, markup it has the potential to be whatever the user wants and as simple/complex as _they_ need.

Searchable, readable universal text files that can organize all. In a grep-accessible place. That can have one or multiple graphical interfaces to tackle any/multiple of these features. Accessible on all your devices. In a future-proof way. Without depending on third-party services.

I have been looking for years and through many many solutions. It doesn't _have_ to be org-mode, but I mean what are the alternatives with that description?

**Comment by zerodat** [comment]
@9z0  I think we are on the same page.  I looked for an org-mode desktop app, didn't find one, started to write one, and then sleek lured me back to todo.txt.  I do like outliners (check out [Vimflowy](https://www.wuthejeff.com/vimflowy/#) for instance).   Org-mode seemed attractive, but it is hard to implement fully, and I'm not sure I need all of my todos and notes combined into one document.  So I have instead built separate solutions for note-taking and just want something clean and simple for keeping track of tasks.  It sounds like you want the two more integrated together, maybe something like [TaskPaper – Plain text to-do lists for Mac](https://www.taskpaper.com/).  This type of thing does appeal to me, but I haven't ever found just the right thing. Definitely want a plain-text system, or at least one where I control my own data and don't need the cloud.

Also there is the platform issue.  For me, it needs to be an open-source linux app, preferably with a compatible open-source Android app that I can keep in sync.  With sleek and [simpletask-android](https://github.com/mpcjanssen/simpletask-android), I have that combination for todo.txt.

**Comment by 9z0** [comment]
>@9z0 I think we are on the same page.

👍

---

>I do like outliners (check out [Vimflowy](https://www.wuthejeff.com/vimflowy/#) for instance). 

>[TaskPaper – Plain text to-do lists for Mac](https://www.taskpaper.com/)

Thanks. Yeah, although they both seem limited to one line notes.

---

> and I'm not sure I need all of my todos and notes combined into one document.

> It sounds like you want the two more integrated together

Not necessarily! I don't like cramming everything together either. You can have a multiple GUI apps or just one or two, but at the end, your todo file(s) and note file(s) are accessible in text files. You get the benefits of visual organization and the benefits of text formats.

It is about applying the benefit of todo.txt + Sleek to actual notes rather than one-line todos.

Sometimes notes about a topic can have multiple aspects and so if you want organization using plain text, notes tend to scatter around in multiple places, and you end up with duplicates/old information, trying to remember all the time (where did I put it? What words did I use so I can search? etc) . One the other hand, a large text file in it self is difficult to manage and keep clear.

I always find myself trying to organize the mess going into circles of:

splitting things into folders and files for clarity > "where should I put this?" > "where did I put this?" > search notes folders using different terms > "I need to organize things better in one place" > A relational database > difficult and slow to search and filter without a custom-made UI > don't know how to make one > "Nothing developed by others fits my needs" > "This one is very good and I can adapt." > "I doesn't have a mobile app!" > Try org-mode or something similar > Difficult to filter and manage in plain text and Orgzly doesn't have that important feature > Try a non-standard mobile app for a portion of notes/todos important to have on-the-go. > The app doesn't have a desktop equivalent!

And now I have my notes in 6 different places and 8 different formats!

**Comment by zerodat** [comment]
> I always find myself trying to organize the mess going into circles of:

> splitting things into folders and files for clarity > "where should I put this?" > "where did I put this?" > search notes folders using different terms > "I need to organize things better in one place" > A relational database > difficult and slow to search and filter without a custom-made UI > don't know how to make one > "Nothing developed by others fits my needs" > "This one is very good and I can adapt." > "I doesn't have a mobile app!" > Try org-mode or something similar > Difficult to filter and manage in plain text and Orgzly doesn't have that important feature > Try a non-standard mobile app for a portion of notes/todos important to have on-the-go. > The app doesn't have a desktop equivalent!

> And now I have my notes in 6 different places and 8 different formats!

Yes, you have described my history perfectly.  I have used most every todo system out there, and many many note-taking systems, and never could find one to commit to for more than a year or two.  The cloud solutions always failed me at some point due to unexpected code changes and instability, and of course, there is the privacy issue as well.  So I have committed to self-hosted and local solutions, synchronized with syncthing, hosted on my own network.  But still there are too many choices and not enough solid functionality that is supported on all of my platforms.  Currently I use a wiki-like system for notes in Markdown, which I edit in my editor and then generate a locally hosted searchable static site with hugo or similar static site generator.  I'm hoping that Markdown will prove to be a format that I can continue to use for a very long time, because of the diverse software options, like zettlr, stackedit, dillinger, bear on the mac, etc etc.  I don't actually use any of these at the moment, but having the option to move my data to them without format changes is very attractive.  Editing notes in my code editor is a good thing because of all the muscle memory I have for the shortcuts and features of that editor (vscodium).

So bringing this back to sleek, if someone were to write a good desktop GUI app that supported the same org-mode subset as orgzly does on Android, I would be interested.  But we seem to agree that org-mode would be difficult to support alongside todo.txt in sleek itself.
